File: modules/module-provider-sdk/src/main/scala/io/constellation/provider/sdk/ModuleDefinition.scala
package io.constellation.provider.sdk

import cats.effect.IO

import io.constellation.provider.TypeSchemaConverter
import io.constellation.provider.v1.provider as pb
import io.constellation.{CType, CValue}

/** Describes a module that a provider offers for execution.
  *
  * @param name
  *   Short module name (e.g., "analyze")
  * @param inputType
  *   CType for the module's input
  * @param outputType
  *   CType for the module's output
  * @param version
  *   Semantic version string (informational)
  * @param description
  *   Human-readable description
  * @param handler
  *   The function that executes this module
  */
final case class ModuleDefinition(
    name: String,
    inputType: CType,
    outputType: CType,
    version: String,
    description: String,
    handler: CValue => IO[CValue],
    batchHandler: Option[List[CValue] => IO[List[Either[Throwable, CValue]]]] = None,
    maxBatchSize: Int = 0
) {

  /** Whether this module supports batch execution (RFC-034 Phase 1). */
  def supportsBatch: Boolean = batchHandler.isDefined

  /** Convert to a protobuf ModuleDeclaration for registration. */
  def toDeclaration: pb.ModuleDeclaration =
    pb.ModuleDeclaration(
      name = name,
      inputSchema = Some(TypeSchemaConverter.toTypeSchema(inputType)),
      outputSchema = Some(TypeSchemaConverter.toTypeSchema(outputType)),
      version = version,
      description = description,
      capabilities = Some(
        pb.ModuleCapabilities(
          supportsBatch = supportsBatch,
          maxBatchSize = maxBatchSize
        )
      )
    )

  /** Produce the fully qualified module name within a namespace. */
  def qualifiedName(namespace: String): String = s"$namespace.$name"
}

File: sdks/typescript/src/types/module-definition.ts
/**
 * Describes a module that a provider offers for execution.
 *
 * Mirrors the Scala `ModuleDefinition` case class.
 */

import type { CType } from "./ctype.js";
import type { CValue } from "./cvalue.js";
import { TypeSchemaConverter } from "../serialization/type-schema-converter.js";
import type { ModuleDeclaration } from "../transport/transport.js";

export interface ModuleDefinition {
  /** Short module name (e.g., "analyze"). */
  name: string;
  /** CType for the module's input. */
  inputType: CType;
  /** CType for the module's output. */
  outputType: CType;
  /** Semantic version string (informational). */
  version: string;
  /** Human-readable description. */
  description: string;
  /** The async function that executes this module. */
  handler: (input: CValue) => Promise<CValue>;
}

/** Convert a ModuleDefinition to a protobuf ModuleDeclaration for registration. */
export function toDeclaration(mod: ModuleDefinition): ModuleDeclaration {
  return {
    name: mod.name,
    inputSchema: TypeSchemaConverter.toTypeSchema(mod.inputType),
    outputSchema: TypeSchemaConverter.toTypeSchema(mod.outputType),
    version: mod.version,
    description: mod.description,
  };
}

/** Produce the fully qualified module name within a namespace. */
export function qualifiedName(mod: ModuleDefinition, namespace: string): string {
  return `${namespace}.${mod.name}`;
}
